### ISSUE TITLE
Deleting and unregistering constraints on EP after generating EP through generate_model.jl

### ISSUE THREAD

**Issue by kalebsm** [GenXProject/GenX#307]
Hello,

I am working on generating a model with the "generate_model" function and then subsequently updating specific constraints of that model. I don’t want to edit the functions of “generate_model.jl”  and my hope is to delete and unregister the constraint and write a new constraint posthumously. I want to keep intact the generate_model process so that I can benchmark my changes against the original. 

I'll use a registered constraint as an example though; after generating the model, there is a constraint "registered" as cSoCBalStart. After running `generate_model`, trying: `is_valid(EP, cSoCBalStart)` returns an error `UndefVarError: cSoCBalStart not defined`. 

For some reason, `unregister(EP, cSoCBalStart)` seems to work. cSoCBalStart is a ::symbol in JuMP, but I feel I should still be able to validate and delete it regardless.

Also, I checked the JuMP documentation and issues and found a related issue (https://github.com/jump-dev/JuMP.jl/issues/1956), but I did not find it to lead me to the solution I was hoping for.

Do you have any idea what I can do to posthumously delete and unregister constraints in the EP model?

Relatedly, some constraints are not registered while others are. Is there a specific reason for this? 

Thanks,
Kaleb

**Comment by RuaridhMacd** [comment]
Hi @kalebsm. you should be able to do this but it's tricky to diagnose. I'm not 100% sure, but I don't think unregister! will fail even if there's nothing to unregister as long as the variable name exists. 

If cSoCBalStart is a specific concern, could you attach your generators_data file and settings folder?

**Comment by kalebsm** [comment]
@RuaridhMacd From the documentation here it says that the object still needs to be deleted, though (https://jump.dev/JuMP.jl/stable/reference/models/#JuMP.unregister). Does keeping that object mean the constraint is still there, or does unregistering the constraint essentially delete it from the model (ie EP)? And cSoCBalStart was just a simple example I chose because it is one of the constraints that is registered. The constraints I really care about are the Minimum Up and Down times. Currently, I am just doing some tests on the ISONE One Zone example.

@sambuddhac I am trying work on a loop of short 48 hour models where the parameters of the constraint would change depending on the loop I was in. Instead of going through each layer of the code from "my" new run.jl to generate_model.jl and thermal_commit.jl, I was hoping to just make changes in the run.jl file I am creating. (this is on a commit before the run.jl file was changed to call run_genx_case, by the way.)

Thanks.

**Comment by RuaridhMacd** [comment]
@kalebsm You're right that you need to delete and unregister the constraint. Just doing one or the other will cause problems.

Looking at the min up time and down constraints, it looks like they're sometime anonymous constraints (e.g. lines 231 and 237 of thermal_commit.jl), which could make this approach difficult.

As a workaround, you could create a run file including a loop which runs GenX, changes the minimum up and down times in your input data, then rebuilds a fresh JuMP model. It's not as efficient as your approach but making a fresh model each time is safer and it should be easier to test out.

I've attached an example of this approach here. I added the .txt suffix to let me upload it here, so you'll have to rename it. It loops over a set of investment variables for one of my technologies. It's a little older so please compare it to the case_runners in the repo to make sure it's taking all the necessary steps when running GenX.

[Run_loop.jl.txt](https://github.com/GenXProject/GenX/files/10026908/Run_loop.jl.txt)

**Comment by kalebsm** [comment]
Hello, I have found the solution to deleting containerized constraints. This is not in the documentation but delete has to be broadcasted, so the command is `delete.(EP, 'constraint')`

Since most constraints in GenX are containerized this may be useful to someone in the future. 

Also, I found that deleting constraints is time-consuming and should be avoided in loops. I think @RuaridhMacd your workaround would be a good approach to take. There are some useful tools described in this discourse (https://discourse.julialang.org/t/deleting-containerized-constraints/21971/5). 

I think I can close this close this issue now. Thanks all for the help.
